src/tasks/authenticate/taskpasswordauth.cpp

```diff
@@ -0,0 +1,246 @@
+
+#include "taskpasswordauth.h"
+#include "hkdf.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+#include "networkmanager.h"
+
+#include <QCryptographicHash>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QPasswordDigestor>
+#include <QRandomGenerator>
+#include <QUrl>
+#include <QUrlQuery>
+
+namespace {
+Logger logger(LOG_MAIN, "TaskPasswordAuth");
+}
+
+TaskPasswordAuth::TaskPasswordAuth(const QString& username,
+                                   const QString& password)
+    : Task("TaskPasswordAuth"), m_username(username) {
+  MVPN_COUNT_CTOR(TaskPasswordAuth);
+
+  // Process the user's password into an FxA auth token
+  QString salt =
+      QString("identity.mozilla.com/picl/v1/quickStretch:%1").arg(username);
+  QByteArray pbkdf = QPasswordDigestor::deriveKeyPbkdf2(
+      QCryptographicHash::Sha256, password.toUtf8(), salt.toUtf8(), 1000, 32);
+
+  HKDF hash(QCryptographicHash::Sha256);
+  hash.addData(pbkdf);
+  m_authpw = hash.result(32, "identity.mozilla.com/picl/v1/authPW");
+}
+
+TaskPasswordAuth::~TaskPasswordAuth() { MVPN_COUNT_DTOR(TaskPasswordAuth); }
+
+void TaskPasswordAuth::run(MozillaVPN* vpn) {
+  Q_UNUSED(vpn);
+
+  QRandomGenerator* generator = QRandomGenerator::system();
+  Q_ASSERT(generator);
+
+  // Generate a PCKE verification code.
+  static QByteArray range(
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~");
+  m_verifier.clear();
+  for (uint16_t i = 0; i < 128; ++i) {
+    m_verifier.append(range.at(generator->generate() % range.length()));
+  }
+
+  // Generate the PKCE challenge.
+  QByteArray challenge =
+      QCryptographicHash::hash(m_verifier, QCryptographicHash::Sha256);
+
+  QUrlQuery query;
+  query.addQueryItem("code_challenge", challenge.toBase64());
+  query.addQueryItem("code_challenge_method", "S256");
+  query.addQueryItem("user_agent", NetworkManager::userAgent());
+
+  // Login using the android auth flow to convince the API to supply the OAuth
+  // code as a query parameter during redirection instead of invoking a browser
+  QUrl url = NetworkRequest::apiBaseUrl();
+  url.setPath("/api/v2/vpn/login/android");
+  url.setQuery(query);
+
+  NetworkRequest* request =
+      NetworkRequest::createForGetUrl(this, url.toString(), 200);
+
+  connect(request, &NetworkRequest::requestFailed, this,
+          &TaskPasswordAuth::netRequestFailed);
+  connect(request, &NetworkRequest::requestHeaderReceived, this,
+          &TaskPasswordAuth::vpnLoginHeaders);
+}
+
+void TaskPasswordAuth::authFailed(ErrorHandler::ErrorType error) {
+  logger.log() << "Authentication failed" << error;
+  MozillaVPN::instance()->errorHandle(ErrorHandler::AuthenticationError);
+  emit completed();
+}
+
+void TaskPasswordAuth::netRequestFailed(QNetworkReply::NetworkError error,
+                                        const QByteArray& data) {
+  Q_UNUSED(data);
+  authFailed(ErrorHandler::toErrorType(error));
```

Comment:
In theory here we should check if the email validation is required.
See: errno=125 + verificationEmail = email-captcha

When this happens we should send a request to: /v1/account/login/send_unblock_code then continue passing the code.